### PR TITLE
Expose new public api for rate limiting and user blocking

### PR DESCRIPTION
**Pull request by timokoessler**
## Example for express
```ts
import Zen from "@aikidosec/firewall";

app.use((req, res, next) => {
    Zen.setUser({
        id: "123",
        name: "John Doe",
    });
    next();
});

Zen.setupExpressIntegration(app);

```

## Generic
```ts
import Zen from "@aikidosec/firewall";

app.use(async (req, res) => {
    Zen.setUser({
        id: "123",
        name: "John Doe",
    });
    
    const result = Zen.shouldBlockRequest();
    if (result.block) {
        if (result.type === "rate-limit") {
            // ...
        }
    }
});
```

### CODE REVIEW

File: library/integrations/express.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * Attacks will still be blocked by Zen if you do not call this function.
+ * Execute this function as early as possible in your Express app, but after the middleware that sets the user.
+ */
+export function setupExpressIntegration(app: Express) {
+  app.use((req, res, next) => {
+    const result = shouldBlockRequest();
+
+    if (result.block) {
+      if (result.type === "rate-limit") {
+        let message = "You are rate limited by Aikido firewall.";
```
Comment (hansott) [inline library/integrations/express.ts:16]:
```suggestion
        let message = "You are rate limited by Zen.";
```

File: library/integrations/hono.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * Attacks will still be blocked by Zen if you do not call this function.
+ * Execute this function as early as possible in your Hono app, but after the middleware that sets the user.
+ */
+export function setupHonoIntegration(app: Hono) {
+  app.use(async (c, next) => {
+    const result = shouldBlockRequest();
+
+    if (result.block) {
+      if (result.type === "rate-limit") {
+        let message = "You are rate limited by Aikido firewall.";
```
Comment (hansott) [inline library/integrations/hono.ts:16]:
```suggestion
        let message = "You are rate limited by Zen.";
```

File: library/integrations/hono.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * Calling this function will setup rate limiting and user blocking for the provided Hono app.
+ * Attacks will still be blocked by Zen if you do not call this function.
+ * Execute this function as early as possible in your Hono app, but after the middleware that sets the user.
+ */
+export function setupHonoIntegration(app: Hono) {
+  app.use(async (c, next) => {
+    const result = shouldBlockRequest();
+
+    if (result.block) {
+      if (result.type === "rate-limit") {
```
Comment (hansott) [inline library/integrations/hono.ts:15]:
```suggestion
      if (result.type === "rate-limited") {
```

or

```suggestion
      if (result.type === "ratelimited") {
```

Comment (hansott) [reply library/integrations/hono.ts:15]: Or at least align with `blocked`? Both past tense or both current?

File: docs/express.md [hunk 4]
```diff
@@ -20,22 +20,32 @@ import "@aikidosec/firewall";
...
 It's recommended to enable this on your staging environment for a considerable amount of time before enabling it on your production environment (e.g. one week).
 
+## Rate limiting and user blocking
+
+If you want to add the rate limiting feature to your app, modify your code like this:
+
+```js
+const Zen = require("@aikidosec/firewall");
+
+const app = express();
```
Comment (hansott) [inline docs/express.md:42]: Let's include `setUser(...)` in an example middleware?

Ad a route under `Zen.setupExpressIntegration` to show the order?

File: docs/hapi.md [hunk 5]
```diff
@@ -23,22 +23,32 @@ import "@aikidosec/firewall";
...
+If you want to add the rate limiting feature to your app, modify your code like this:
+
+```js
+const Zen = require("@aikidosec/firewall");
+
+const server = Hapi.server(...);
+
+// Call this as early as possible, before other onRequest extensions are added
+// If you want to use user based rate limiting or blocking, you need to call Zen.setUser(...) in a onRequest function before this
+Zen.setupHapiIntegration(app);
```
Comment (hansott) [inline docs/hapi.md:49]: Same here

File: docs/hono.md [hunk 6]
```diff
@@ -22,22 +22,32 @@ import "@aikidosec/firewall";
...
+If you want to add the rate limiting feature to your app, modify your code like this:
+
+```js
+const Zen = require("@aikidosec/firewall");
+
+const app = Hono();
+
+// Call this as early as possible, before other middleware
+// If you want to use user based rate limiting or blocking, you need to call Zen.setUser(...) in a middleware before this
+Zen.setupHonoIntegration(app);
```
Comment (hansott) [inline docs/hono.md:48]: Same here

File: library/index.ts [hunk 7]
```diff
@@ -7,3 +8,17 @@ const shouldEnable = shouldEnableFirewall();
...
+import { setupHonoIntegration } from "./integrations/hono";
+import { setupHapiIntegration } from "./integrations/hapi";
+
+export {
+  setUser,
+  shouldBlockRequest,
+  setupExpressIntegration,
+  setupHonoIntegration,
+  setupHapiIntegration,
+};
```
Comment (hansott) [inline library/index.ts:24]: So on Deno it will crash but only after logging the message, right?

File: library/integrations/express.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+ */
+export function setupExpressIntegration(app: Express) {
+  app.use((req, res, next) => {
+    const result = shouldBlockRequest();
+
+    if (result.block) {
+      if (result.type === "ratelimited") {
+        let message = "You are rate limited by Zen.";
+        if (result.trigger === "ip" && result.ip) {
+          message += ` (Your IP: ${result.ip})`;
```
Comment (hansott) [inline library/integrations/express.ts:17]: escape? just to be sure

File: library/integrations/hapi.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * Execute this function as early as possible in your Hapi app, but after the middleware that sets the user.
+ */
+export function setupHapiIntegration(app: Server) {
+  app.ext("onRequest", function onRequest(request, h) {
+    const result = shouldBlockRequest();
+    if (result.block) {
+      if (result.type === "ratelimited") {
+        let message = "You are rate limited by Zen.";
+        if (result.trigger === "ip" && result.ip) {
+          message += ` (Your IP: ${result.ip})`;
```
Comment (hansott) [inline library/integrations/hapi.ts:16]: escape? just to be sure

File: library/integrations/hono.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ */
+export function setupHonoIntegration(app: Hono) {
+  app.use(async (c, next) => {
+    const result = shouldBlockRequest();
+
+    if (result.block) {
+      if (result.type === "ratelimited") {
+        let message = "You are rate limited by Zen.";
+        if (result.trigger === "ip" && result.ip) {
+          message += ` (Your IP: ${result.ip})`;
```
Comment (hansott) [inline library/integrations/hono.ts:17]: escape? just to be sure

File: library/integrations/shouldBlockRequest.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    return { block: true, type: "blocked", trigger: "user" };
+  }
+
+  const rateLimitResult = shouldRateLimitRequest(context, agent);
+  if (rateLimitResult.block) {
+    return {
+      block: true,
+      type: "ratelimited",
+      trigger: rateLimitResult.trigger,
+      ip: context.remoteAddress ? escapeHTML(context.remoteAddress) : undefined,
```
Comment (hansott) [inline library/integrations/shouldBlockRequest.ts:32]: this feels indeed the safest to do but if you want to output the response as JSON in an API... Then unescaped would be better? See other comments

File: library/integrations/express.ts (hunk 8 above)
Comment (timokoessler) [reply library/integrations/express.ts:17]: The ip is already escaped inside the shouldBlockRequest function

File: library/integrations/shouldBlockRequest.ts (hunk 11 above)
Comment (timokoessler) [reply library/integrations/shouldBlockRequest.ts:32]: In both cases it's not a valid ip inside the json, so does it matter? This protects users who use `shoudBlockRequest` directly and output it to html.

File: library/index.ts (hunk 7 above)
Comment (timokoessler) [reply library/index.ts:24]: With new Deno v2:
```
error: Uncaught (in promise) ReferenceError: exports is not defined
Object.defineProperty(exports, "__esModule", { value: true });
                      ^
    at file:///Users/timokoessler/Git/firewall-node/build/index.js:2:23

    info: Deno supports CommonJS modules in .cjs files, or when there's a package.json
          with "type": "commonjs" option and --unstable-detect-cjs flag is used.
    hint: Rewrite this module to ESM,
          or change the file extension to .cjs,
          or add package.json next to the file with "type": "commonjs" option
          and pass --unstable-detect-cjs flag.
    hint: See https://docs.deno.com/go/commonjs for details
```

File: docs/generic-middleware.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+- The framework must be based on the Node.js http / https module (this includes most frameworks).
+- The package `@aikidosec/firewall` must be included in your app before the framework is loaded or the http server is created.
+
+## Example
+
+```js
+const Zen = require("@aikidosec/firewall");
+
+function onRequest(...){
```
Comment (hansott) [inline docs/generic-middleware.md:15]:
```suggestion
function onRequest(...) {
```

File: test.ts [hunk 13]
```diff
@@ -0,0 +1,3 @@
+import "./build/index.js";
+
+console.log("test");
```
Comment (hansott) [inline test.ts:3]: Remove file?

**Review by hansott — APPROVED** [review]

**Pull request merged**